tests/jdom/tester.ts

```diff
@@ -0,0 +1,193 @@
+import { JDBus } from "../../src/jdom/bus"
+import JDServiceProvider from "../../src/jdom/serviceprovider"
+import { EVENT, DEVICE_ANNOUNCE } from "../../src/jdom/constants";
+import { mkBus } from "../testutils";
+
+import { JDEvent } from "../../src/jdom/event";
+import { JDDevice } from "../../src/jdom/device";
+import RoleManager from "../../src/servers/rolemanager"
+import JDServiceServer from "../../src/jdom/serviceserver";
+import { assert } from "../../src/jdom/utils";
+import { JDService } from "../../src/jdom/service";
+
+
+// how the heck is this not a native operation
+function setEquals<T>(set1: Set<T>, set2: Set<T>): boolean {
+    if (set1.size != set2.size) {
+        return false
+    }
+    set1.forEach(value => {
+        if (!set2.has(value)) {
+            return false
+        }
+    })
+    return true
+}
+
+
+interface ServerDevice {
+    server: JDServiceServer,
+    roleName?: string,
+}
+
+
+// Creates a bus with the specified servers (optionally bound to the specified roles),
+// then runs the test function.
+export async function withBus(devices: ServerDevice[], 
+    test: (bus: JDBus, serviceMap: Map<JDServiceServer, JDService>) => Promise<void>) {
+    const bus = mkBus()
+
+    // For server devices: add the service provider on the bus and return the device
+    const serverDevices = devices.map(device => {
+        const busDevice = bus.addServiceProvider(new JDServiceProvider(
+            [(<ServerDevice>device).server] 
```

Comment:
Do you need the cast?

Reply:
Nope, removed. Was leftover code when removing out the trace device support.